### PR TITLE
2.x: Add Observable switchMapX and concatMapX operators

### PR DESCRIPTION
**Pull request by akarnokd**
This PR ports the various `switchMapX` and `concatMapX` operators from #5870, #5871, #5872 and #5873 to `Observable`:

- `concatMapCompletable` (replaced by the common implementation)
- `concatMapCompletableDelayError`
- `concatMapMaybe`
- `concatMapMaybeDelayError`
- `concatMapSingle`
- `concatMapSingleDelayError`
- `switchMapCompletable`
- `switchMapCompletableDelayError`
- `switchMapMaybe`
- `switchMapMaybeDelayError`
- `switchMapSingle` (dedicated implementation)
- `switchMapSingleDelayError` (dedicated implementation)

This PR concludes the requested set of operators in #4853.

Marbles will be updated/adjusted in a separate PR.

### CODE REVIEW

File: src/main/java/io/reactivex/internal/operators/mixed/ObservableConcatMapCompletable.java [hunk 1]
```diff
@@ -0,0 +1,277 @@
...
+        }
+
+        @Override
+        public void onError(Throwable t) {
+            if (errors.addThrowable(t)) {
+                if (errorMode == ErrorMode.IMMEDIATE) {
+                    disposed = true;
+                    inner.dispose();
+                    t = errors.terminate();
+                    if (t != ExceptionHelper.TERMINATED) {
```
Comment (vanniktech) [inline src/main/java/io/reactivex/internal/operators/mixed/ObservableConcatMapCompletable.java:123]: this path does not seem to be covered. same for 126. Is travis just flaky again?

File: src/main/java/io/reactivex/internal/operators/mixed/ObservableSwitchMapCompletable.java [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+
+        @Override
+        public void onError(Throwable t) {
+            if (errors.addThrowable(t)) {
+                if (delayErrors) {
+                    onComplete();
+                } else {
+                    disposeInner();
+                    Throwable ex = errors.terminate();
+                    if (ex != ExceptionHelper.TERMINATED) {
```
Comment (vanniktech) [inline src/main/java/io/reactivex/internal/operators/mixed/ObservableSwitchMapCompletable.java:130]: this one too

File: src/main/java/io/reactivex/internal/operators/mixed/ObservableSwitchMapSingle.java [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+
+        @Override
+        @SuppressWarnings({ "unchecked", "rawtypes" })
+        public void onNext(T t) {
+            SwitchMapSingleObserver<R> current = inner.get();
+            if (current != null) {
+                current.dispose();
+            }
+
+            SingleSource<? extends R> ms;
```
Comment (vanniktech) [inline src/main/java/io/reactivex/internal/operators/mixed/ObservableSwitchMapSingle.java]: should be `ss`

File: src/test/java/io/reactivex/internal/operators/mixed/ObservableConcatMapCompletableTest.java [hunk 4]
```diff
@@ -0,0 +1,362 @@
...
+        cs.onError(new TestException());
+
+        assertFalse(pp.hasObservers());
+
+        to.assertFailure(TestException.class);
+    }
+
+    @Test
+    public void boundaryError() {
+        PublishSubject<Integer> pp = PublishSubject.create();
```
Comment (vanniktech) [inline src/test/java/io/reactivex/internal/operators/mixed/ObservableConcatMapCompletableTest.java]: `ps`

**Review by vanniktech — APPROVED** [review]
Some small nits. Maybe those renames could be done in a separate PR since this one is already super huge.

File: src/main/java/io/reactivex/internal/operators/mixed/ObservableConcatMapCompletable.java (hunk 1 above)
Comment (akarnokd) [reply src/main/java/io/reactivex/internal/operators/mixed/ObservableConcatMapCompletable.java:123]: That's a probabilistic race that the CI tends to not hit, no matter the race-loop count. It gets covered 80% of the time and unfortunately, I'd have to hack the state of the operator to get it execute that inner part a second time.

Comment (vanniktech) [reply src/main/java/io/reactivex/internal/operators/mixed/ObservableConcatMapCompletable.java:123]: Yeah thought so. Then never mind.

**Review by vanniktech — APPROVED** [review]

**Comment by akarnokd** [comment]
Reopening due to the Travis/GitHub service problems.  The build succeeded but the PR doesn't seem to pick up that state.

**Review by vanniktech — APPROVED** [review]
Now it's all good

**Pull request merged**